Fix #11659, problems with tab characters not counted correctly

ScottPJones opened the pull request:
This fixes issues with tabs not being counted correctly (they were counted as a fixed with of 8 spaces,
instead of moving over to the next tab stop), in the `Base.indentation` and `Base.unindent` functions,
which are used to copy/paste text in `REPL.jl` and `LineEdit.jl`.
This also allows code to set the tab width (`Base.tab_indentation_width`).
This also adds unit tests so that the functions should be completely covered.

ScottPJones committed:
Fix #11659, tab characters counted incorrectly in indentation and unindent

Fix REPL test

pao commented:
Did you mean to include 225eb89 in here (also included in #11718 as ca1d5c3)?

ScottPJones commented:
I was just about to push a change to remove that dependency, as I found the code in test/repl.jl that was broken...

pao commented:
rgr, thanks

ScottPJones commented:
I know everybody here *loves* terseness, but sometimes, it hinders understanding... what does rgr mean, please? :grinning:

pao commented:
"roger"...sorry, aviation background...

ScottPJones commented:
OK, I just pushed the version without the #11718 change, now that the test is fixed here.
I was afraid that not understanding the `rgr` was a generational thing, with all of the younger genius hotshots in this community! :grinning:

ScottPJones commented:
Grrrr!  What now is broken on Appveyor and Travis?  Again, it doesn't seem to be related to my change... and passed once on each...

pao commented:
Only failure appears to be Travis 32-bit; these tests were last modified March 7 https://github.com/JuliaLang/julia/blame/master/test/ccall.jl#L121

cc @tkelman is this a new failure mode (sorry to bother you, but you've become the domain expert)?

tkelman commented:
Some broken stuff got committed to master, that was the commit being merged into. I restarted the build.

tkelman commented on base/string.jl:
```diff
@@ -1007,38 +1011,71 @@ function indentation(s::AbstractString)
+                col = div(col + tabwid, tabwid) * tabwid
+            elseif ch == '\n'
+                # Now we need to output enough indentation
+                for i = 1:max(0, col-indent) ; write(buf, ' ') ; end
```
A for-loop over an empty range will get skipped, so the `max` shouldn't be necessary here.
Stylistically I don't think 1-line for loops are all that readable. There was one here already which you're replacing, but you're also adding 3 more.

ScottPJones replied:
I was just following the style of the code I was replacing, but I can make those 3 lines.

ScottPJones commented:
@tkelman I updated the for loops

tkelman commented:
And the other part of my comment? You should be able to replace `for i = 1:max(0, col-indent)` with `for i = 1:(col-indent)`

ScottPJones commented:
@tkelman Sorry! Just missed that in the rush to fix it, all done now.

StefanKarpinski commented on test/repl.jl:
```diff
@@ -177,29 +203,6 @@ begin
-    \tls
-    # time: 2014-06-30 20:44:29 EDT
-    # mode: julia
-    \t2 + 2"""
```
I really don't think the unindent code should be changing the indentation of these lines at all. The behavior of this was intended to be that if and only if the last line in the double quotes are pure white space, is that whitespace stripped from each preceding line of the quoted data. Perhaps that's not what got implemented though.

ScottPJones replied:
This all comes from @nolta's change, https://github.com/JuliaLang/julia/commit/f8b3f9e0ad730bed588227f476993ad021241943
If the last line is not empty after stripping, it walks through every non-blank line, and figures out the minimum indentation and uses that for unindent.
It looks like a lot of code depends on that behavior also.

StefanKarpinski replied:
Yeah, that's unfortunate if I'm understanding this situation correctly. @nolta, how did you intend for this to work?

ScottPJones replied:
Given that that *is* the state of affairs, and code already depends on the behavior that @nolta added, do you see a problem with my change here?

StefanKarpinski replied:
Yes, that's true – this code is about creating fake REPL history data with the right content, which this does.

nolta replied:
#2305

ScottPJones commented:
That's not at all what the code did.

ScottPJones commented:
Bump!  Any reason why this can't be merged?  Thanks!

StefanKarpinski commented:
So I've said it before and I'll say it again – I really don't think we shouldn't be assuming anything about the width of tabs. The unindent function should strip a common whitespace *string* from the beginning of indented text, not a indent width. If the leading whitespace isn't consistent, it should be an error.

ScottPJones commented:
? This change doesn't change anything about the width of tabs, it does *calculate* the tab stop correctly, unlike before, and allows the width of the tabs to be changed from the default (which other people asked for).
With this change, you get good results from the `unindent` function, unlike before.
The `unindent` function doesn't take a whitespace *string* to match, it takes a position.
What you are suggesting can't even be done with the current function.

StefanKarpinski commented:
Right, that's what I was trying to convey when you set out to do this – if you're even making assumptions about tab sizes, you're doing it wrong (not *you* specifically, but the generic "you"). This is definitely a better implementation of `unindent`, but it should take the tab size as an argument (maybe a keyword arg), which I guess can default to the global value. Since the standard indent size for Julia is 4 rather than 8, I would argue that should be the default for that. But the bigger picture is that we should change the triple quote unindent so that it doesn't make any assumptions about tab size at all.

ScottPJones commented:
Making assumptions about tab sizes is wrong?   That's part of many standards, as I pointed out earlier!
You are *supposed* to use those values (or defaults) if you are going to deal with any of them.
Having the default tab size overrideable in `unindent` is a good idea though.  I can add that easily.
Should it just be positional? With the default picking up the global value?

Tab size is *not* equivalent to indent size... that is where people have been getting messed up for years... (bad editors that didn't have smart tab and delete, and then allowed people to set the tab width... yechhh!).  If you want something to display nicely everywhere, you really should stick to tabs meaning move over to the next mod 8 position... and use only spaces, or tabs followed by spaces, for whatever indentation you feel like.

Stripping indentation from triple-quote is really separate from what a general `unindent` or `dedent` function should do, but I still don't see that ignoring standards that define what the default tab positions should be is good.

StefanKarpinski commented:
The standard tab size for Julia code is 4 spaces.

mbauman commented:
Am I missing something here?  `indentation` and `unindent` are unexported internal functions that seem to only serve one purpose: removing a common whitespace prefix from every line.  For that purpose, it can be tab-size agnostic.  So why not be tab-size agnostic?

StefanKarpinski commented:
Yes, exactly my point. Having an unident utility function somewhere (a [text formatting package](https://github.com/carlobaldassi/TextWrap.jl) maybe), and that should certainly handle tabs correctly, but that's irrelevant to Base Julia which simply shouldn't be assuming things about tabs at all.

ScottPJones commented:
The problem is, when the indentation to be removed is less than the tab size, when not all lines have identical whitespace at the beginning of each line, which would be a common occurrence for people you prefer to use tabs and not just spaces in their programs.

StefanKarpinski commented:
If any of the lines in an indented triple-quoted block doesn't start with the exact whitespace sequence that the last line does, it should be an error. This is how Python does it and it's the only way to be sure that the programmer is getting what they think they're getting.

ScottPJones commented:
@mbauman Maybe if there had been some internal documentation about the intended purpose of these functions, from the beginning, this code wouldn't have been changed in a way that the original author didn't intend (and I'm not talking about this PR either...)
If this code really was supposed to be tab-agnostic, and remove a common whitespace prefix from every line, then *that* is what it should have done, instead of incorrectly calculating the indentation, using a fixed width for each tab character, and an `unindent` function that simply check for a common whitespace prefix, because it's only passed a number (incorrectly calculated, to boot).

Also, and this is a general *big* problem with Julia, saying a function is not exported doesn't really mean anything, because anybody can simply call it by doing `Base.unindent`.  But that's a topic for another day.

StefanKarpinski commented:
Thank you, as usual, for your diagnosis of our failures at large.

ScottPJones commented:
@StefanKarpinski OK, then I'd say the following... let this PR go in to fix the `indentation` and `unindent` functions, which very well may be used by people, exported or not, and I can make another PR that has
`stripprefix`, that can take a line and an arbitrary prefix to strip, which can then be used to make triple quote tab agnostic.

>Thank you, as usual, for your diagnosis of our failures at large.

Come on.  You'd better grow a thicker skin, if you are going to do language design.
Am I *wrong* in my diagnosis's?  That's the only thing that matters.

Also, I've said time and again, Julia is the most promising language I've seen in ages, and that all the contributors should be proud of their accomplishments.
I wouldn't spend a second trying to help resolve the problems I do see in it, if I didn't feel that the language had so much potential.

However, does that mean that it is perfect? No.  Should everybody stay silent over any flaws they see, major or minor, because they might hurt the feelings of the contributors? No.  My hope is that any flaws can be corrected in the very short term, before the user community gets even larger... I don't think it would be good to see a Python 2 vs Python 3 sort of split in the future.  I hope that "Arraymeggedon" (and a "Stringnarok" :grinning:) can happen quickly, this year, while there is time, before julia gets too big.

mbauman commented:
I've never touched this code myself, but it wasn't hard to grep through `base/` and see where it is being used.  There are only two uses: pasting at the REPL and triple-quoted strings.  I think a rename could make sense here, but there's definitely no reason to support these unexported functions for users that are inadvisably using them.  There's enough other work to be done.

> Am I wrong in my diagnosis's? That's the only thing that matters.

I strongly disagree.  We don't want to be alienating other collaborators by calling their work *wrong* or characterizing it as having big problems.  Some may have a thick enough skin to handle it, but I don't want anyone to get bullied away.  Suggest fixes, and point out areas for improvement, please.  But don't harp on them.

StefanKarpinski commented:
My skin is plenty thick. The annoying thing about these frequent commentaries about what you perceive to be the shortcomings of the project is that they are usually off-topic, not constructive, and no longer relevant to the project. Case in point: you are complaining about the lack of documentation of an internal utility function that was introduced over two years ago. Shall we build a time machine and go back in time to enforce the @ScottPJones way on past contributors? I'll start working on the flux capacitor!

There's no point in these side remarks about this sort of thing. What is actually constructive is figuring out how things out to behave. To that end, +1 to the `stripprefix` plan. I'm not sure we need to keep `unindent` around – it's not exported or documented, so anyone who relies on it is asking for breakage.

ScottPJones commented:
@mbauman I've been doing nothing but suggest fixes and point out areas for improvement!
Not just suggest, but spend a significant amount of my own time trying to fix things myself.
I've added documentation, fixed issues in the scheme parser, fixed issues in the `utf8proc` library,
improved the `@time` macros, fixed a number of Unicode related bugs.
I've also tried to get people to see the value of internal documentation (and adding it myself) and the value of a language helping you keep your abstraction layer and implementation layer separate... with loads of suggestions (and code).
I don't think that counts as "harping" on them.
If I see something as a major problem, which I do about issues of abstraction vs. implementation, should I just downplay that?

@StefanKarpinski Come on, I'm not saying that you can change the past, but I'm not seeing any change in behavior in the present.  That is what concerns me.  It's not the "Scott Jones" way, it's just good software engineering practices (which I learned mostly from Gerry Sussman and Barbara Liskov), that I think would be *helpful* for Julia's development in the future.

mbauman commented:
You don't need to defend your contributions here.  They've been valuable and I try to help you out where I can. I am simply saying that your tone matters.

You don't need to list all the ways `indention` and `unindent` are doing things "wrong".  Just help us make it better.

ScottPJones commented:
>You don't need to list all the ways indention and unindent are doing things "wrong". Just help us make it better.

@mbauman How do you describe what is fixed, unless you say what was wrong before the fix?

StefanKarpinski commented:
Nobody has any issue with the fact that the way the `unindent` function currently works is wrong. Identifying that is totally ok. Talking about general problems with the project as a whole while trying to figure out how to fix a fairly specific thing is at best distracting and irrelevant, and on the hundredth time, pretty annoying. Anyway, let's stop talking about this and just fix the thing.

ScottPJones commented:
Also, I *am* just trying to help make things better!
I think this is a good fix for `indentation` and `unindent`, to make things better.
You also didn't look at the package source, and I did find a few occurrences of `Base.unindent` and
`Base.indentation`.

>If any of the lines in an indented triple-quoted block doesn't start with the exact whitespace sequence that the last line does, it should be an error. This is how Python does it and it's the only way to be sure that the programmer is getting what they think they're getting.

@StefanKarpinski You should double check Python... that's not true at all.

ScottPJones commented:
You may be getting confused with what Python's help system does with docstrings.
Triplequoted strings are raw strings in Python, including new-lines and all, they don't dedent at all.

ScottPJones commented:
>Talking about general problems with the project as a whole while trying to figure out how to fix a fairly specific thing is at best distracting and irrelevant,

It's not irrelevant at all.  I was trying to point out, that this is yet another example of *why* current behavior should be changed... both in having people start doing internal documentation, and in having some facility to be able to keep functions like these from being used in packages (same thing with all the abuse of .data on strings).

johnmyleswhite commented:
@ScottPJones: You need to think more about the social implications of your behavior. Technical correctness **is** irrelevant if people dislike working with you.

ScottPJones commented:
Ufff... so, better not to bring up *technical issues*, in case it might make people dislike working with me?
Have I really been impolite to people here?   Have I made personal and derogatory comments to anybody?   AFAICR, I've only commented on *code*.  Other people have decided to attack me personally.

StefanKarpinski commented:
If you truly believe that your behavior is innocuous, why have so many people reacted so negatively to it? This is a pretty friendly, unrancorous community – no one has ever drawn the amount of fire you have. Please consider reflecting on this and maybe internalizing a bit of the feedback you're getting.

ScottPJones commented:
So, just what is it that makes people react negatively?  I've never called anybody a dick, for example.

StefanKarpinski commented:
There has been ample fairly specific feedback, I'm not going to repeat it here. Usually you just argue with it; you may want to use it for self-improvement instead.

mbauman commented:
Here's how this got derailed: up until https://github.com/JuliaLang/julia/pull/11719#issuecomment-113572406 we were considering if this could be written in a tab-size agnostic manner.  That comment was a shotgun of "you're doing it wrongs:" internal documentation, the way things should have been, and private methods. The issues with this specific code had already been identified (and belong in #11659), and the larger issues are way out of scope.  There's no need to say those things in response to a potential solution.

ScottPJones commented:
@mbauman I felt it was relevant to point out where this particular code had become problematical... it was difficult to change because external code was calling it, and it had problems because changes had been made, not in accordance with the original author's wishes, because of a lack of internal documentation.  How are people going to realize that there's a problem, unless they are given *specific* examples of why there's a problem?

ScottPJones commented:
@StefanKarpinski The feedback I got, was to not be so verbose, or to talk so much about my experience.
What else?  I've tried to improve, both in that, and in technical things, like making smaller PRs.

mbauman commented:
> How are people going to realize that there's a problem

Ah, I think this may be a big part of it.  *Your audience is often already aware of the problems*.  So repeatedly pointing them out comes across as abrasive and annoying.

kshyatt commented:
@ScottPJones For a long time, I've had a tendency to interpret questions the other party meant innocuously, e.g. "Katie, did you know that in C++11 you can inherit base class constructors?", in a semi-insulting way so that what I *heard* was "Katie, I'm implying you didn't know this *super obvious* fact, probably because you're stupid!". Of course rationally I know that's not what they meant, but sometimes it's easy for me to "jump the gun" and respond in a kind of snippy way. When I read https://github.com/JuliaLang/julia/issues/11749#issuecomment-113253167, it reminded me of this pattern that I sometimes have. I've started quite a few internet fights by doing this - it's hard to interpret tone through text.

ScottPJones commented:
@kshyatt Well, after somebody has gone as far as calling you a dick, and never even apologized, then I think that interpreting the tone of such a question from that person as derogatory is fairly natural.

I've never meant to annoy *anybody*.  I've praised julia and the people who've created it to the skies, so that even my 9 year old son is trying to learn julia.  I just don't understand why then, people constantly think *I* am attacking.

@mbauman Maybe if I saw some recognition that the audience was aware of all the problems, that would be a different story... when I've brought things up, mostly I've got back that things are just fine as they are... it's very frustrating.

kshyatt commented:
@ScottPJones I agree that calling someone a "dick" is of course not neutral and derogatory. I interpreted your question as "why did people react negatively to me initially, and why did it keep escalating?"

ScottPJones commented:
I'd love to know the answer to that... I've been trying to take any constructive criticism to heart , both technical and personal, and have been profusely thankful for that help...

kshyatt commented:
I don't mean to speak for or over anyone else here, but here's my impressions of what happened:
1. You submitted some big PRs with the technical issues that have been discussed quite a bit
2. People reacted with various criticisms/questions, and imo, the pattern I mentioned before happened 
3. You and a few others got caught in a mutually-reinforcing cycle of being annoyed at each other and less charitable than either side might otherwise have been
4. Steps 1-3 repeated a few times, so now both "sides" are a bit entrenched.

I believe you when you say "I've been trying to take any constructive criticism to heart". I've noticed definite improvements in your PR and commenting style :). But change doesn't happen overnight, and like I said above, I've had what I think is a similar issue for a long time (I'm still working on it). In the space between figuring out what the behavior is that needs improvement, and being 100% done with fixing it, I've managed to annoy quite a few people, even though they were aware I was trying to work on it. I hope my comment was helpful to you.

ScottPJones commented:
About 3.  I really haven't been annoyed at all at any of the technical debate... to me, that's where you can really learn, and the outcome (hopefully), is that you end up convinced of the other person(s) argument(s), or they end up convinced of yours', or you all end up finding a much better solution that either on their own would not have come up with.   I thought I'd expressed my thankfulness for even the harshest reviews (as long as it was constructive).  I learned soooo very much in the process.

I hope in our few interactions, you didn't feel I was ever critical of *you*... (I've been incredibly impressed by all the bug fixing, etc. you've managed to do in such a short time, and I just saw today you are even giving a talk next week).

Your comments very helpful, thank you!

kshyatt commented:
As I said, it's just my impression, but I think others might agree that some of the comment threads have become heated (on both sides) whether that's being annoyed at the technical debate or just annoyed at each other.

>I hope in our few interactions, you didn't feel I was ever critical of you... (I've been incredibly impressed by all the bug fixing, etc. you've managed to do in such a short time, and I just saw today you are even giving a talk next week).

Thank you! No, I haven't felt unfairly criticized at all. You asked for feedback, and as someone who hasn't really participated in these debates much (I don't know a single thing about Unicode!), I thought you might like an "outside" view.

ScottPJones commented:
Technical things frustrate me, if they don't get fixed... but I'm really a pretty easy going person, and the only thing that has annoyed me has been the personal derogatory remarks (as opposed to constructive criticism, like about the length of my posts, which I have tried to improve on).
It still surprises me, that people think I'm attacking their work, because I was sure that at least the people who met me at Jeff's thesis defense would have seen me as an unmitigated fanboy, of julia and her creators and contributors.

kshyatt commented:
>Technical things frustrate me, if they don't get fixed

I think this is universally true among programmers :)

>It still surprises me, that people think I'm attacking their work, because I was sure that at least the people who met me at Jeff's thesis defense would have seen me as an unmitigated fanboy, of julia and her creators and contributors.

I think Jeff and Stefan (and possibly others?) have explained what specifically is driving this feeling, but I don't want to speak for them.

Also, just fyi, I think per the [Julia Community Standards](http://julialang.org/community/standards/),
> In particular, do not sexualize the term “Julia” or any other aspects of the project. While “Julia” is a female name in many parts of the world, the programming language is not a person and does not have a gender.

so I believe the preferred pronoun for the Julia PL is "it".

mbauman commented:
This is awesome.  :)

Scott, I'm really happy that you've persevered and are still here!

One last tip - you don't need to respond personally to every comment on PRs and issues, even yours. Remember that you're not just having a one-on-one conversation (or multiple interleaved one-on-one conversations), but it's more akin to sitting on a stage discussing the issues in front of a large audience. In general, I try to limit the one-on-one back and forth, and instead I try to keep the thread as a whole coherent and moving forward.

ScottPJones commented:
OK, folks!  Have at it... this leaves the fixes for `indentation` and `unindent` in place, as I saw them used in packages, and revamps triplequote so that it doesn't depend on things like `RevString`, `search`,`IOBuffer`,`print`,`write` etc... as well as fixing the bugs by using the new function `stripindent!` I also added, which *is* tab size agnostic, unlike the old triplequote code.

KristofferC commented:
When it comes to what tone your text gives off I believe this is a good text to read, especially for you @ScottPJones : https://sonatano1.wordpress.com/2013/12/23/how-to-misuse-your-ellipses-and-infuriate-your-readers/

StefanKarpinski commented:
Please let's not bring up tone right now – things are calm and everyone is coding productively and collaborating nicely at the moment.

ScottPJones commented:
It would be nice if people gave a heads up if they decide to start working on an issue that somebody else was already fixing. Case in point, #11815.  I haven't looked at that implementation yet, and I do think it is better to have it in the parser, but not warning someone actively working on an issue, letting them waste their time, doesn't count as collaborating nicely.
This PR still fixes the `indentation` and `unindent` functions, which are used in `LineEdit` and `REPL` in Base, and possibly elsewhere in packages.

ivarne commented:
I don't really think there is a good way to coordinate a project like Julia to avoid double work like this. Everybody have different goals and need to prioritize their time between fixing bugs, implementing features, answering questions, and reviewing code. Someone (preferably you; if you had searched for [triplequoted](https://github.com/JuliaLang/julia/search?q=Triplequoted+&type=Issues&utf8=✓)) should have referenced #2682 from your issue or PR, but now that it didn't happen, we should move on with a single mention of the wasted effort (not two). 

I'd even argue that it wasn't such a waste, because you were better prepared for reviewing @nolta's PR, so that it can be merged faster.

ScottPJones commented:
@ivarne - it wasn't a case of @nolta not knowing I was working on this (I had pointed out his change that caused the problem 6 days ago, and he had responded).  I didn't reference #2682, because this change doesn't do anything about that problem.  A simple: "Hey, I think I'm going to move handing triplequoting into the parser to fix this" would have sufficed, to stop me from wasting any time this weekend on it.
Since he was definitely informed that I was working on this, to me that seems just common courtesy.

Also, I was already prepared to review @nolta's PR, from my first pass at fixing the bugs in his f8b3f9e change from 2 years ago (which has no PR associated with it that I could find), so it was just a total waste of time.

StefanKarpinski commented:
Pull requests are experiments. Some of them work, some of them don't. Looking at a pull request that didn't get merged as a waste of time is not the right perspective – it was an experiment that didn't pan out, which happens all the time in any experimental endeavor. As long as the most successful experiment wins out in the end, everyone wins.

ScottPJones commented:
@StefanKarpinski That doesn't get me back the time I wasted.  If I'd known that @nolta was going to start tackling handling this in the parser, I would not have spent more time on this PR.  If that were the case, I'd be fine with it.  That's all I'm saying.  (Also, I think the first commit, not the second, should still be merged... I'll remove the second commit).

ScottPJones commented:
I've pulled out the changes to `triplequoted`, since those are being superseded by #11815 (which is a better way of handing it, and also fixes another issue), so that this just has the fixes to `unindent` and `indentation`, which are used in at least two places in Base, and also in packages)

ScottPJones commented:
What is going on *now* with Appveyor and Travis???  It's dying with some C or C++ declarations now.
Not mine!

tkelman commented:
This needs a rebase. Travis was an OOM killer (that still happens once in a while), appveyor was broken the night before juliacon because of libstdc++ ABI compatibility issues (llvm-config didn't run correctly, so all llvm include flags were missing). After you rebase a win64 timeout is likely to happen due to #11818

ScottPJones commented:
All rebased

ScottPJones commented:
This failed on AppVeyor in spawn.  I don't think that's mine.

ScottPJones commented:
Again, 2 of the 3 Travis builds failed due to totally unrelated changes.
This has been rebased again (which is why it is picking up the new Travis failures, unfortunately).

ScottPJones committed:
Add coverage testing for indentation/unindent functions

ScottPJones commented:
This is mergeable again, and has added some more unit testing, so that all of the modified code should be completely covered (the old code which this PR fixes had 0 tests).

ScottPJones commented:
Bump: this should fix the bug and bring string/io.jl to 100% coverage (the untested part represents about 25% of the total)

tkelman commented:
Since the change of how triple quotes get parsed, where are `indentation` and `unindent` still used?

ScottPJones commented:
In REPL.jl and LineEdit.jl

tkelman commented:
Hm, looks like that originates from https://github.com/Keno/REPL.jl/issues/35. How important is it that we do whatever those lines of code are doing in bracketed paste mode? If the indentation is a little odd for multi-line code pasted at the repl would anyone notice? Seems like we could try taking that out and just deprecating or removing these, doesn't look like any registered packages are using them.

ScottPJones commented:
Why would you want to leave broken, totally untested, uncovered code in Julia, when there is a working fix, with full coverage tests, already available?

tkelman commented:
Come on man, that's not what I want at all. The technical issues here from https://github.com/JuliaLang/julia/pull/11719#issuecomment-113530081 about assuming a hard-coded tab size remain. So even if we merge this PR, and it's certainly an improvement so we may as well, I'm saying that this particular code serves very little purpose and we could probably get rid of it at very little cost.

ScottPJones commented:
OK, I just think that, if you want a different approach for handling the unindentation in REPL.jl and LineEdit.jl, that can be done at some later point, instead of deprecating or removing stuff, which would be more effort than just merging this.  (besides, I thought you didn't like simply removing stuff, even stuff that is not used in any registered package?)
Also, I paste in multi-line stuff frequently, so I'd notice.

tkelman commented:
I don't like simply removing exported things without any deprecation period, especially not late in a release cycle or when that deletion is included along with other larger changes. Deprecation of exported things is the right way to move towards deleting them, which is a good thing when they're not widely used or doing so helps achieve other important goals. Deprecation of unexported internal functions that aren't widely used is more controversial, as you've seen, and may or may not be necessary.

So if this helps with pasting multi-line code, can you come up with an example where this PR changes the behavior in a way that's an obvious improvement?

ScottPJones commented:
Demonstrably correct code (with fully coverage from unit tests) is not an obvious improvement?
I've always believed in improving code quality, whether or not external behavior is changed.

tkelman commented:
I'm talking about actual end-user visible behavior in multi-line paste, which is the only place this gets used. 100% line coverage is great, but it doesn't mean there's zero chance this could be introducing any new bugs. I'm not saying we shouldn't merge this, I'm saying this particular piece of code is not worth this much conversation or effort.

ScottPJones commented:
External behavior is definitely fixed by this PR, I just didn't know how to write a unit tests for multi-line paste, just did it manually and saw the bad results before, and the correct results now.
```julia
         @inbounds while out < len
        	ch::UInt32 = dat[pos += 1]
             # Handle ASCII characters
```
vs.
```julia
         @inbounds while out < len
             ch::UInt32 = dat[pos += 1]
             # Handle ASCII characters
```
Having everything turn really ragged when pasted just didn't seem nice to me.

tkelman commented:
Great, that's all that I asked for. Interacting with you has been extremely frustrating, and it's not been improving. Fewer people over time have continued to attempt doing it.

I'm going to make the tab width a keyword argument to `indent` and `unindent` instead of being a global. globals are bad.

ScottPJones commented:
>Interacting with you has been extremely frustrating, and it's not been improving.

If you have any suggestions as to what you feel I'm doing wrong, please let me know
(preferable by e-mail).
I am trying to learn how to interact better with all the ins-and-outs of an OSS community, and
appreciate any and all constructive criticism.

>I'm going to make the tab width a keyword argument to indent and unindent instead of being a global. globals are bad.

If you'd brought that up at any time in the review, I would have been happy to change it.
You are the only person who said anything about that, after 88 comments.

tkelman commented:
https://github.com/JuliaLang/julia/pull/11719#issuecomment-113541417

ScottPJones commented:
OK, I missed the part about the keyword argument a month ago, however he didn't say that the global should be removed:

>which I guess can default to the global value
